Sources/Stevenson/Codable Utils/CustomCodable.swift

```diff
@@ -0,0 +1,30 @@
+import Foundation
+
+public protocol CodableTransformer {
+    associatedtype DecodedType
+    associatedtype TypeForCoding: Codable
+    static func encode(decodedObject: DecodedType) throws -> TypeForCoding
+    static func decode(encodedObject: TypeForCoding) throws -> DecodedType
+}
+
+@propertyWrapper
+public struct CustomCodable<Transformer: CodableTransformer> {
+    public var wrappedValue: Transformer.DecodedType
+    public init(wrappedValue: Transformer.DecodedType) {
+        self.wrappedValue = wrappedValue
+    }
+}
```

Comment:
This `@protocolWrapper` looks very nice!

My two cents for slight improvement is in https://gist.github.com/inamiy/eb7112bb943d8d2ed4dc3fffce4055a0 .
We can remove `TypeForCoding: Codable` constraint to make a protocol more generalised, and instead constrain on `extension` so that generalised protocol can be used in many other scenarios.

But it's OK to keep this code as is 👍

Reply:
That's a nice idea to make the protocol be usable for other transformations besides `Codable`… but tbh I'm not a fan of the naming in your gist 😅 #NamingIsHard `convert(to: To) -> From` reads a bit strange for example given that it reads "convert to X" but in fact converts from a `To` to a `From`, feels a bit backwards on reading 🤔 

But if we find better names for that (not an easy task 😜) it's a good idea to decouple that in the future 👍

Reply:
Some more ideas: `func convert / invert`, `forward / backward` , `fromTo / toFrom` ...

Reply:
```swift
@CustomCodable(.YMD) let date: Date

struct CustomCodable<A, B> {
    let iso: PartialIso<A, B>
    init(_ iso: PartialIso<A, B>) { ... }
}

extension PartialISO where A == String, B == Date {
  static var YMD: PartialISO {
     /* date formatter here */
  }
}
```
😉 
https://github.com/pointfreeco/swift-web/blob/f3946a64ea5a529e7023f4bf0f4e9c82d991bc82/Sources/ApplicativeRouter/PartialIso.swift

Reply:
@ilyapuchka it's gonna work for `Codable` but not `Decodable`.
Because when decoding, we need to `init` the enclosing type passing both the `wrappedValue` (which we'd fetch from the JSON)… and the `iso` property – which we can't, since we're in the middle of `init` so we need that `.YMD` property to be `static` to be able to use it, the value of `iso` won't be provided by the JSON

That's why it's doable for `Codable` (which already has an instance that you encode) but not for `Decodable` (for which you're in the process of creating the instance so you don't have `iso`. Which led me to that solution of using `static` property and generics.

Reply:
Yeah I also got stuck to use `self.iso` inside `self.init(from decoder: Decoder)` because of error: `Constant 'self.iso' used before being initialized` .

BTW alongside pointfree's impl, it seems `apply` / `unapply` is quite common naming to be used.

- https://github.com/schernichkin/partial-isomorphisms/blob/master/src/Control/Isomorphism/Partial/Prim.hs
- https://github.com/paulyoung/purescript-partial-isomorphisms/blob/v1.0.0/src/Control/Isomorphism/Partial.purs

But it might not work well if we change `Optional` return type to more accurate `Result` or `Either` to carry Error type as well.

Reply:
Oh I like `apply/unapply`. Good use it for when I share that PW trick in my blog post 😛